File: Signiture-Backend/delivery/api/src/main/java/pcy/study/api/common/api/ApiResponse.java
package pcy.study.api.common.api;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Builder;
import org.springframework.http.ResponseEntity;

@Builder
public record ApiResponse<T>(
        @Schema(description = "결과")
        ApiResult result,
        @Schema(description = "데이터")
        T body
) {

    public static <T> ResponseEntity<ApiResponse<T>> ok(T data) {
        return ResponseEntity.ok(create(ApiResult.ok(), data));
    }

    public static ResponseEntity<ApiResponse<Void>> error(ApiCode apiCode) {
        return new ResponseEntity<>(create(ApiResult.error(apiCode)), apiCode.getHttpStatus());
    }

    public static ResponseEntity<ApiResponse<Void>> error(ApiCode apiCode, String message) {
        return new ResponseEntity<>(create(ApiResult.error(apiCode.getCode(), message)), apiCode.getHttpStatus());
    }

    private static ApiResponse<Void> create(ApiResult result) {
        return create(result, null);
    }

    private static <T> ApiResponse<T> create(ApiResult result, T data) {
        return ApiResponse.<T>builder()
                .result(result)
                .body(data)
                .build();
    }
}

File: Signiture-Backend/kotlin/delivery/common/src/main/kotlin/pcy/study/common/api/ApiResponse.kt
package pcy.study.common.api

import jakarta.validation.Valid

data class ApiResponse<T>(
    var result: ApiResult? = null,
    @field:Valid
    var body: T? = null
) {

    companion object {

        @JvmStatic
        fun <T> ok(data: T): ApiResponse<T> {
            return ApiResponse(ApiResult.ok(), data)
        }

        @JvmStatic
        fun error(apiCode: ApiCode): ApiResponse<Unit> {
            return ApiResponse(ApiResult.error(apiCode))
        }

        @JvmStatic
        fun error(
            apiCode: ApiCode,
            message: String
        ): ApiResponse<Unit> {
            return ApiResponse(ApiResult.error(apiCode.getCode(), message))
        }
    }
}